### PR TITLE
Incremental builds

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds incremental builds and snapshots to bake. This allows bake to only rebuild targets based on the state of the dependent files and targets.

/cc @titanous

### CODE REVIEW

File: snapshot.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+		return 0, err
+	}
+
+	// If it's a regular file then hash the mtime.
+	if fi.IsDir() {
+		return 0, nil
+	}
+
+	// Generate hash from file contents.
+	h := fnv.New64a()
```
Comment (titanous) [inline snapshot.go]: This should be a SHA-256.

**Comment by benbjohnson** [comment]
@titanous Changed to SHA-256 in 94897ef. You'll need to `rm -rf ~/.bake` to rebuild.

File: cmd/bake/main.go [hunk 2]
```diff
@@ -104,81 +150,83 @@ func (m *Main) Run() error {
...
+		if err != nil {
+			return err
+		}
+		defer os.Chdir(wd)
+
+		// Change to new directory
+		if err := os.Chdir(path); err != nil {
+			return err
+		}
+	*/
```
Comment (titanous) [inline cmd/bake/main.go]: Commented code?

File: snapshot.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+// Directories are hashed using its mode and a list of the directory's files.
+func hashFileInfo(path string) (uint64, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0, err
+	}
+
+	// If it's a regular file then hash the mode, mtime, & size.
+	if !fi.IsDir() {
+		h := fnv.New64a()
```
Comment (titanous) [inline snapshot.go]: I think we should be using SHA-256 everywhere, no reason to risk collisions with a tiny non-cryptographic hash.

File: cmd/bake/main.go (hunk 2 above)
Comment (benbjohnson) [inline cmd/bake/main.go]: The pwd was moved to the `exec.Cmd` so it wasn't needed here. I forgot to pull out the commented section. Fixed in 9fae1c3.

File: snapshot.go (hunk 3 above)
Comment (benbjohnson) [inline snapshot.go]: Converted to SHA-256 in 9fae1c3.

**Pull request merged**